### ISSUE TITLE
[PHP]heavily access to roots.pem

### ISSUE THREAD

**Issue by axot** [grpc/grpc#18291]
I found there are heavily io access to `etc/roots.pem`, this will cause high sys usage especially when using overlay file system.

```
4828522 10:05:14.321589480 19 php-fpm (16054) < open fd=8(<f>/project/vendor/grpc/grpc/etc/roots.pem) name=/project/vendor/grpc/grpc/etc/roots.pem flags=1(O_RDONLY) mode=0
4828523 10:05:14.321591000 19 php-fpm (16054) > fstat fd=8(<f>/project/vendor/grpc/grpc/etc/roots.pem)
4828524 10:05:14.321591572 19 php-fpm (16054) < fstat res=0
4828525 10:05:14.321592088 19 php-fpm (16054) > lseek fd=8(<f>/project/vendor/grpc/grpc/etc/roots.pem) offset=0 whence=1(SEEK_CUR)
4828526 10:05:14.321592625 19 php-fpm (16054) < lseek res=0
4828527 10:05:14.321593384 19 php-fpm (16054) > fstat fd=8(<f>/project/vendor/grpc/grpc/etc/roots.pem)
4828528 10:05:14.321593835 19 php-fpm (16054) < fstat res=0
4828529 10:05:14.321594757 19 php-fpm (16054) > read fd=8(<f>/project/vendor/grpc/grpc/etc/roots.pem) size=8192
```

**Comment by axot** [comment]
And also /project/vendor/grpc/grpc/composer.json. Is it nessary?

**Comment by axot** [comment]
One idea is load these resource into opcache, any idea?

**Comment by axot** [comment]
hello @stanley-cheung, could you give some advice please?

**Comment by axot** [comment]
Hello @stanley-cheung , any updates for this issue?

**Comment by axot** [comment]
Opacahe will be a simple solution, the problem is how to generate roots.pem.php. Any idea? I could try to push a merge request.

**Comment by stanley-cheung** [comment]
@axot, how did you enable those logs in php-fpm?

**Comment by axot** [comment]
Hello @stanley-cheung , thank you for handling this issue, we are using https://github.com/draios/sysdig to investigate our Kuberenets pods. You could run it with the following command inside GKE.

sudo docker run -it --rm --name sysdig --privileged --network host -e SYSDIG_BPF_PROBE="" -v /var/run/docker.sock:/host/var/run/docker.sock -v /dev:/host/dev -v /proc:/host/proc:ro -v /boot:/host/boot:ro -v /lib/modules:/host/lib/modules:ro -v /usr:/host/usr:ro -v /etc:/host/etc:ro --entrypoint /bin/bash sysdig/sysdig:0.24.2

The log is recorded using csysdig's `dig` function. You can also check `file I/O` which also shows roots.pem access.

**Comment by castaneai** [comment]
Hi, I'm also interested in this problem.

When I looked inside PHP, the following line read the file:
https://github.com/grpc/grpc/blob/master/src/php/lib/Grpc/BaseStub.php#L47

I am using Spanner Client through php-fpm, and `Grpc::BaseStub::__construct` is called four times per request.

**Comment by dwsupplee** [comment]
@castaneai the main Cloud Spanner client wraps around 4 generated clients:

- A client to interface with long running operations
- Database admin
- Instance admin
- Spanner data

As of `google/cloud-spanner` [v1.15.1](https://github.com/googleapis/google-cloud-php/releases/tag/v0.100.0) we have begun lazily instantiating the long running/admin clients instead of configuring them up front. With this latest version, do you observe less IO resources being utilized?

**Comment by axot** [comment]
@dwsupplee I inspect the file i/o today with `google/cloud-spanner` v1.15.1 which get 60MB/s and 10K  OPS, in the previous version it almost take 250MB/s. Now we confirmed less io for roots.pem but may need more effort, any thoughts?

**Comment by dwsupplee** [comment]
Your proposal of loading the roots file into opcache seems reasonable to me, and as you alluded would require some updates to the toolchain to make sure the generated rooms.pem.php file is kept up to date. @stanley-cheung, are you open to this proposal as well?

**Comment by stanley-cheung** [comment]
@dwsupplee Yes we are open to trying to load the roots file into opcache to avoid repeated IO. Actually I have assigned someone to look into this a while ago but just haven't made it happen.

**Comment by stanley-cheung** [comment]
Just to close on this - this fix is included in the `1.26.0` release (currently in release candidate phase as of now).

The fix is that `ChannelCredentials::setDefaultRootsPem()` will only be called once for the life of the `grpc` extension being loaded, instead of repeatedly on the same `$ssl_roots` file. Hence `file_get_contents()` should only be called once per process.

I ran some benchmarks and confirmed the behavior. I started a PHP client and ran a script which makes 1 unary request, and measure the time it takes to request this .php script 1000 times. 

Prior to the fix (1.25.0), 1000 requests take 14.699s on average.
After the fix (1.26.0), 1000 requests take 14.348s on average.
The difference is 351ms for 1000 requests.

I also took some measurements on just 1 `file_get_contents($path)` call, which takes 277us, on average.  So that, very roughly, explains the 351ms that we saved after the fix, per 1000 requests.
